core/src/raw/oio/write/block_write.rs

```diff
@@ -0,0 +1,180 @@
+/// uploads. By implementing BlockWrite, services don't need to
+/// care about the details of uploading blocks.
+///
+/// # Architecture
+///
+/// The architecture after adopting [`BlockWrite`]:
+///
+/// - Services impl `BlockWrite`
+/// - `BlockWriter` impl `Write`
+/// - Expose `BlockWriter` as `Accessor::Writer`
+///
+/// # Notes
+///
+/// `BlockWrite` has an oneshot optimization when `write` has been called only once:
+///
+/// ```no_build
+/// w.write(bs).await?;
+/// w.close().await?;
+/// ```
+///
+/// We will use `write_once` instead of starting a new block upload.
+/// We will use `write_once` instead of starting a new block upload.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait BlockWrite: Send + Sync + Unpin + 'static {
+    /// write_once is used to write the data to underlying storage at once.
+    ///
+    /// BlockWriter will call this API when:
+    ///
+    /// - All the data has been written to the buffer and we can perform the upload at once.
+    async fn write_once(&self, size: u64, body: AsyncBody) -> Result<()>;
+
+    /// write_block will write a part of the data and returns the result
+    /// [`Block`].
+    ///
+    /// BlockWriter will call this API and stores the result in
+    /// order.
+    ///
+    /// - block_number is the index of the block, starting from 0.
+    async fn write_part(&self, size: u64, block_number: usize, body: AsyncBody) -> Result<Block>;
+
+    /// complete_block will complete the block upload to build the final
+    /// file.
+    async fn complete_block(&self, parts: &[Block]) -> Result<()>;
+
+    /// abort_block will cancel the block upload and purge all data.
+    async fn abort_block(&self, parts: &[Block]) -> Result<()>;
+}
+
+/// The result of [`BlockWrite::write_block`].
+///
+/// services implement should convert Block to their own represents.
+///
+/// - `block_id` is the id of the block.
+#[derive(Clone)]
+pub struct Block {
+    /// The id of the block.
+    pub block_id: String,
+    /// block_number is the index of the block, starting from 0.
+    pub block_number: usize,
+}
+
+struct WriteBlockFuture(BoxedFuture<Result<Block>>);
+
+/// # Safety
+///
+/// wasm32 is a special target that we only have one event-loop for this WriteBlockFuture.
+unsafe impl Send for WriteBlockFuture {}
+
+/// # Safety
+///
+/// We will only take `&mut Self` reference for WriteBlockFuture.
+unsafe impl Sync for WriteBlockFuture {}
+
+impl Future for WriteBlockFuture {
+    type Output = Result<Block>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.get_mut().0.poll_unpin(cx)
+    }
+}
+
+/// BlockWriter will implements [`Write`] based on block
+/// uploads.
+pub struct BlockWriter<W: BlockWrite> {
+    state: State,
+    w: Arc<W>,
+
+    parts: Vec<Block>,
+    cache: Option<oio::ChunkedBytes>,
+    futures: ConcurrentFutures<WriteBlockFuture>,
+    next_block_number: usize,
```

Comment:
block itself doesn't have order, the order is decided by the `blocks_id` in `complete_block`. We can use `uuid` to generate block id.

Reply:
How do we guarantee the order of `block_ids` in `complete_block`？

Reply:
> How do we guarantee the order of `block_ids` in `complete_block`？

The `blocks: Vec<String>` must be ordered.